src/main/kotlin/org/elm/lang/core/resolve/scope/ImportScope.kt

```diff
@@ -1,12 +1,87 @@
 import org.elm.lang.core.psi.ElmNamedElement
 import org.elm.lang.core.psi.elements.ElmImportClause
 import org.elm.lang.core.psi.elements.ElmTypeAliasDeclaration
 import org.elm.lang.core.psi.elements.ElmTypeDeclaration
+import org.elm.lang.core.psi.globalModificationTracker
 import org.elm.lang.core.stubs.index.ElmModulesIndex
 
+private val EXPOSED_VALUES_KEY: Key<ParameterizedCachedValue<ExposedNames, ElmFile>> = Key.create("EXPOSED_VALUES_KEY")
+private val EXPOSED_TYPES_KEY: Key<ParameterizedCachedValue<ExposedNames, ElmFile>> = Key.create("EXPOSED_TYPES_KEY")
+private val EXPOSED_CONSTRUCTORS_KEY: Key<ParameterizedCachedValue<ExposedNames, ElmFile>> = Key.create("EXPOSED_CONSTRUCTORS_KEY")
+
+class ExposedNames(val elements: Array<ElmNamedElement>) {
+    constructor(elements: List<ElmNamedElement>) : this(elements.toTypedArray())
+
+    private val byName = elements.associateByTo(mutableMapOf()) { it.name }.apply { remove(null) }
+    operator fun get(key: String?) = byName[key]
+}
+
+/**
+ * A scope of exposed names for the module named by [qualifierPrefix] reachable
+ * via [clientFile]. By default, the import scopes will be found by crawling the
+ * explicit import declarations in [clientFile] and automatically adding the
+ * implicit imports from Elm's Core standard library.
+ *
+ * @param qualifierPrefix The name of a module or an alias
+ * @param clientFile The Elm file from which the search should be performed
+ * @param importsOnly If true, include only modules reachable via imports (implicit and explicit).
+ *                    Otherwise, include all modules which could be reached by the file's [ElmProject]
+ */
+class QualifiedImportScope(
+        private val qualifierPrefix: String,
+        private val clientFile: ElmFile,
+        private val importsOnly: Boolean = true
+) {
+    fun getExposedValue(name: String): ElmNamedElement? {
+        return scopes().mapNotNull { it.getExposedValues()[name] }.firstOrNull()
+    }
+
+    fun getExposedValues(): Sequence<ElmNamedElement> {
+        return scopes().flatMap { it.getExposedValues().elements.asSequence() }
+    }
+
+    fun getExposedType(name: String): ElmNamedElement? {
+        return scopes().mapNotNull { it.getExposedTypes()[name] }.firstOrNull()
+    }
+
+    fun getExposedTypes(): Sequence<ElmNamedElement> {
+        return scopes().flatMap { it.getExposedTypes().elements.asSequence() }
+    }
+
+    fun getExposedConstructor(name: String): ElmNamedElement? {
+        return scopes().mapNotNull { it.getExposedConstructors()[name] }.firstOrNull()
+    }
+
+    fun getExposedConstructors(): Sequence<ElmNamedElement> {
+        return scopes().flatMap { it.getExposedConstructors().elements.asSequence() }
+    }
+
+    /** Lazily yield all individual scopes that can contain exposed names with this prefix */
+    private fun scopes(): Sequence<ImportScope> = sequence {
+        if (importsOnly) {
+            yieldAll(implicitScopes())
```

Comment:
Shouldn't the explicit scopes be yielded first? I see that this is how the existing code was written, but it's the opposite of what we do in `VisibleNames.getAll()` where first we have top-level decls, then imports, then implicit/global. I haven't been able to produce an example where your way is wrong, but for consistency and sanity, I think it makes sense if explicit comes first.

If you do make this change, we should also change `allScopes` on line 74 to match.